refactor: rework toolchain registration for WORKSPACE and bzlmod users

kormide opened the pull request:
Stop automatically registering all of the toolchains in bazel-lib for WORKSPACE users. Make it easier to override toolchain versions under bzlmod.

Closes https://github.com/aspect-build/bazel-lib/issues/593.

### Type of change

- Refactor (a code change that neither fixes a bug or adds a new feature)

**For changes visible to end-users**

- Breaking change (this change will force users to change their own code or config)
- Relevant documentation has been updated
- Suggested release notes are provided below:
* BREAKING: Toolchains are no longer automatically registered for WORKSPACE users. Users must call the `aspect_bazel_lib_register_toolchains` macro or call the `register_xyz_toolchain` macros individually. See [Release](https://github.com/aspect-build/bazel-lib/releases) for snippets.
* BREAKING: The host repository has been moved from the extension `ext` into a differed module extension named `host`.

### Test plan

- Covered by existing test cases
- New test cases added
- Manual testing; please provide instructions so we can reproduce:

kormide commented:
This is still a WIP, but wouldn't mind some preliminary feedback. Documentation still needs to be updated.

kormide commented on lib/private/extension_utils.bzl:
```diff
@@ -0,0 +1,69 @@
+"""Utility functions for bzlmod extensions"""
+
+def _toolchain_repos(mctx, get_tag, toolchain_name, repos_fn, default_repository = None, get_name = None, get_version = None):
+    """Create toolchain repositories from bzlmod extension tags.
+
+    For uses of the tag class where the name isn't overridden (the `default_repository` name
+    is used), creates a single toolchain repository for the invocation of the tag closest to
+    the root module. Otherwise, resolution follows bzlmod's depth-first search through the
+    module graph.
+
+    Only a root module may declare a toolchain with a name other than the default repository name.
```
This function is confusing so I'm going to do some more iteration on these docs.

kormide commented on lib/private/extension_utils.bzl:
```diff
@@ -0,0 +1,104 @@
+"""Utility functions for bzlmod extensions"""
+
+def _toolchain_repos_bfs(mctx, get_tag_fn, toolchain_name, toolchain_repos_fn, default_repository = None, get_name_fn = None, get_version_fn = None):
```
Calling this one "bfs" because I can envision another strategy to resolve toolchains based on semver.

kormide commented on lib/private/extension_utils.bzl:
```diff
@@ -0,0 +1,104 @@
+            else:
+                registrations[name] = version
+
+        for name, version in registrations.items():
+            toolchain_repos_fn(
+                name = name,
+                version = version,
+            )
```
I'm wondering if this macro is doing too much and we should return the info for which toolchain name/versions to create and not actually create them.

alexeagle reviewed (commented):
I don't think we need to completely stop registering toolchains. Our goal is just to make sure that users don't download unnecessary binaries, and are able to override our toolchain with whatever one they choose, right?

https://github.com/bazelbuild/rules_go/blob/master/docs/go/core/bzlmod.md#registering-go-sdks suggests we should be able to see how they avoid registration taking precedence over users.

alexeagle commented on .github/workflows/release_prep.sh:
```diff
@@ -48,10 +73,24 @@ http_archive(
 )
 
-load("@aspect_bazel_lib//lib:repositories.bzl", "aspect_bazel_lib_dependencies")
+load("//lib:repositories.bzl", "aspect_bazel_lib_dependencies", "register_copy_directory_toolchains", "register_copy_to_directory_toolchains", "register_coreutils_toolchains", "register_expand_template_toolchains", "register_tar_toolchains")
```
in the snippet we should solve for #592 by mentioning which ones you get by default and **don't** have to register.

kormide replied:
It there a reason why we shouldn't just automatically register all of them? They should only be downloaded if used by the corresponding rules, right?

kormide commented:
Updated. Under bzlmod, continue autoregistering toolchains as this does not interfere with the ability for someone to override a toolchain in the root module (see https://bazelbuild.slack.com/archives/C014RARENH0/p1696893626545609). Toolchains can be selectively overridden at different versions.

For WORKSPACE users, added an `aspect_bazel_lib_register_toolchains` macro which can be used to perform all of the registration. Users can optionally register them selectively with the individual toolchain macros.

Made `extension_utils` private while we work out its utility.

alexeagle commented on docs/repositories.md:
```diff
@@ -14,6 +14,21 @@ Load dependencies required by aspect rules
+aspect_bazel_lib_register_toolchains()
+</pre>
+
+Register all bazel-lib at their default versions.
```
all bazel-lib **toolchains**
